compat/socket.cc

```diff
@@ -0,0 +1,194 @@
+    return result;
+}
+
+int
+xlisten(int socketFd, int backlog)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::listen(handle, backlog);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+ssize_t
+xrecv(int socketFd, void * buf, size_t len, int flags)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::recv(handle, static_cast<char *>(buf), static_cast<int>(len), flags);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+ssize_t
+xrecvfrom(int socketFd, void * buf, size_t len, int flags, struct sockaddr * from, socklen_t * fromlen)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    int fl = *fromlen;
+    const auto result = ::recvfrom(handle, static_cast<char *>(buf), static_cast<int>(len), flags, from, &fl);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xsend(int socketFd, const void * buf, size_t len, int flags)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::send(handle, static_cast<const char *>(buf), static_cast<int>(len), flags);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+ssize_t
+xsendto(int socketFd, const void * buf, size_t len, int flags, const struct sockaddr * to, socklen_t l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::sendto(handle, static_cast<const char *>(buf), static_cast<int>(len), flags, to, l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xsetsockopt(int socketFd, int l, int o, const void *v, socklen_t n)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::setsockopt(handle, l, o, static_cast<const char *>(v), n);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xsocket(int domain, int type, int protocol)
+{
+    auto result = ::socket(domain, type, protocol);
+    if (result == INVALID_SOCKET) {
+        SetErrnoFromWsaError();
+        return SOCKET_ERROR;
+    }
+    return _open_osfhandle(result, 0);
```

Comment:
What about setting errno when _open_osfhandle() fails? If setting it is unnecessary here, please add a corresponding C++ comment.

Reply:
sure. EBADF seems like a good pick